rust/lance-index/src/lib.rs

```diff
@@ -14,10 +14,24 @@
 
 //! Lance secondary index library
 
+use std::{any::Any, sync::Arc};
+
+use lance_core::Result;
+
 pub mod scalar;
 pub mod vector;
 
 pub mod pb {
     #![allow(clippy::use_self)]
     include!(concat!(env!("OUT_DIR"), "/lance.index.pb.rs"));
 }
+
+/// Generic methods common across all types of secondary indices
+pub trait Index: Send + Sync {
+    /// Cast to [Any].
+    fn as_any(&self) -> &dyn Any;
```

Comment:
Should we have an enum to identify what Index this is? Downstream user can handle the casting easier

Reply:
We can add that enum.  Right now there is no real downcasting.  The caller either needs a scalar index or they need a vector index and they ask for the index they need.  The only time `open_generic_index` is used is when getting index statistics.